File: src/Wallet.js
export default class Wallet {
    constructor(initialBalance = 0) {
        this.balance = initialBalance;
        this.transactions = [];
    }

    deposit(amount){
        this.balance += amount;
        this.addTransaction(amount,"deposit");
    }

    withdraw(amount) {
        if(this.balance < amount)
            throw "Not enough balance";

        this.balance -= amount;

        this.addTransaction(amount,"withdrawal");
    }

    addTransaction(amount, type) {
        this.transactions.push({"amount":amount, "type":type})
    }
}

File: app/models/wallet.rb
class Wallet < ApplicationRecord
	has_many :transactions
	validates :name, presence: true

	def deposit(amount)
		raise 'Invalid amount' if amount < 0
		self.balance += amount
		Transaction.create({ amount: amount, wallet: self })
		save
	end

	def withdraw(amount)
		raise 'Invalid amount' if amount < 0
		raise 'Insufficient balance' if amount > self.balance
		self.balance -= amount
		Transaction.create({ amount: amount, wallet: self })
		save
	end
end